pkg/util/kubernetes/apiserver/metadata_test.go

```diff
@@ -0,0 +1,443 @@
+
+	stop := make(chan struct{})
+	defer close(stop)
+	informerFactory.Start(stop)
+	go metaController.Run(stop)
+
+	c := client.CoreV1()
+	require.NotNil(t, c)
+
+	// Create a Ready Schedulable node
+	// As we don't have a controller they don't need to have some heartbeat mechanism
+	node := &v1.Node{
+		Spec: v1.NodeSpec{
+			PodCIDR:       "192.168.1.0/24",
+			Unschedulable: false,
+		},
+		Status: v1.NodeStatus{
+			Addresses: []v1.NodeAddress{
+				{
+					Address: "172.31.119.125",
+					Type:    "InternalIP",
+				},
+				{
+					Address: "ip-172-31-119-125.eu-west-1.compute.internal",
+					Type:    "InternalDNS",
+				},
+				{
+					Address: "ip-172-31-119-125.eu-west-1.compute.internal",
+					Type:    "Hostname",
+				},
+			},
+			Conditions: []v1.NodeCondition{
+				{
+					Type:    "Ready",
+					Status:  "True",
+					Reason:  "KubeletReady",
+					Message: "kubelet is posting ready status",
+				},
+			},
+		},
+	}
+	node.Name = "ip-172-31-119-125"
+	_, err := c.Nodes().Create(node)
+	require.NoError(t, err)
+
+	pod := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: metav1.NamespaceDefault,
+		},
+		Spec: v1.PodSpec{
+			NodeName: node.Name,
+			Containers: []v1.Container{
+				{
+					Name:  "nginx",
+					Image: "nginx:latest",
+				},
+			},
+		},
+	}
+	pod.Name = "nginx"
+	pod.Labels = map[string]string{"app": "nginx"}
+	pendingPod, err := c.Pods("default").Create(pod)
+	require.NoError(t, err)
+
+	pendingPod.Status = v1.PodStatus{
+		Phase:  "Running",
+		PodIP:  "172.17.0.1",
+		HostIP: "172.31.119.125",
+		Conditions: []v1.PodCondition{
+			{
+				Type:   "Ready",
+				Status: "True",
+			},
+		},
+		// mark it ready
+		ContainerStatuses: []v1.ContainerStatus{
+			{
+				Name:  "nginx",
+				Ready: true,
+				Image: "nginx:latest",
+				State: v1.ContainerState{Running: &v1.ContainerStateRunning{StartedAt: metav1.Now()}},
+			},
+		},
+	}
+	_, err = c.Pods("default").UpdateStatus(pendingPod)
+	require.NoError(t, err)
+
+	svc := &v1.Service{
+		Spec: v1.ServiceSpec{
+			Selector: map[string]string{
+				"app": "nginx",
+			},
+			Ports: []v1.ServicePort{{Port: 443}},
+		},
+	}
+	svc.Name = "nginx-1"
+	_, err = c.Services("default").Create(svc)
+	require.NoError(t, err)
+
+	ep := &v1.Endpoints{
+		Subsets: []v1.EndpointSubset{
+			{
+				Addresses: []v1.EndpointAddress{
+					{
+						IP:       pendingPod.Status.PodIP,
+						NodeName: &node.Name,
+						TargetRef: &v1.ObjectReference{
+							Kind:      "Pod",
+							Namespace: pendingPod.Namespace,
+							Name:      pendingPod.Name,
+							UID:       pendingPod.UID,
+						},
+					},
+				},
+				Ports: []v1.EndpointPort{
+					{
+						Name:     "https",
+						Port:     443,
+						Protocol: "TCP",
+					},
+				},
+			},
+		},
+	}
+	ep.Name = "nginx-1"
+	_, err = c.Endpoints("default").Create(ep)
+	require.NoError(t, err)
+
+	timeout := time.NewTimer(2 * time.Second)
+
+	// wait for endpoints to be synced by controller
```

Comment:
What happens if we lose connectivity to the API Server ?
I'm assuming the keys are all ejected after the NumRetries as they are processed in the `processNextWorkItem` ?
Do you think we could/should test this out ?

Reply:
This case is already handled by the cache framework. It pretty much just restarts the watch where we left off. See https://github.com/kubernetes/client-go/blob/master/tools/cache/reflector.go#L253 and https://github.com/kubernetes/client-go/blob/master/tools/cache/reflector.go#L132